{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_dna\n",
    "from utils import *\n",
    "\n",
    "import numpy as np\n",
    "import random\n",
    "import pandas as pd\n",
    "\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import GENCODE v34 & construct gene dict\n",
    "\n",
    "Then construct the gene dictionary containing summary of transcript info used for splicing labels: for each gene we find all the transcripts that belong to it, and finding a \"global start\" and \"global end\" as the uttermost left start among all the transcripts and the uttermost right end, respectively, because we will be constructing splicing labels for the whole gene, and we must take a stretch of the sequence that covers all the transcripts of the gene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "227517\n",
      "Took 5.8211989402771 seconds to construct the gene dict\n",
      "Number of genes in the region of interest: 1044\n"
     ]
    }
   ],
   "source": [
    "# 0 - tr. name, 1 - chr, 2 - strand, 3,4 - tr. start/end, 5,6 - exon starts/ends, 7 - gene name\n",
    "transcript_file = np.genfromtxt('./data/GENCODE_v34_hg38_comprehensive', usecols=(1, 2, 3, 4, 5, 9, 10, 12), skip_header=1, dtype='str')\n",
    "print(len(transcript_file))\n",
    "\n",
    "region = ['chr22']\n",
    "\n",
    "gene_dict = {}\n",
    "\n",
    "def mRNA_len(starts, ends):\n",
    "    l = 0\n",
    "    for x in zip(starts, ends):\n",
    "        l += x[1]-x[0]\n",
    "    return l\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "# each key is a gene\n",
    "for row in transcript_file:\n",
    "    \n",
    "    # only for chr22 for now\n",
    "    if row[1] in region:\n",
    "        \n",
    "        if row[7] not in gene_dict.keys():\n",
    "            gene_dict[row[7]] = {}\n",
    "            gene_dict[row[7]]['chr'] = row[1]\n",
    "            gene_dict[row[7]]['strand'] = row[2]\n",
    "            gene_dict[row[7]][row[0][:15]] = {}\n",
    "            gene_dict[row[7]][row[0][:15]]['starts'] = [int(x) for x in row[5].split(',')[:-1]]\n",
    "            gene_dict[row[7]][row[0][:15]]['ends'] = [int(x) for x in row[6].split(',')[:-1]]\n",
    "            gene_dict[row[7]][row[0][:15]]['length'] = mRNA_len(gene_dict[row[7]][row[0][:15]]['starts'],\n",
    "                                                               gene_dict[row[7]][row[0][:15]]['ends'])\n",
    "            gene_dict[row[7]]['global_start'] = int(row[3])\n",
    "            gene_dict[row[7]]['global_end'] = int(row[4])\n",
    "\n",
    "        else:\n",
    "            gene_dict[row[7]][row[0][:15]] = {}\n",
    "            gene_dict[row[7]][row[0][:15]]['starts'] = [int(x) for x in row[5].split(',')[:-1]]\n",
    "            gene_dict[row[7]][row[0][:15]]['ends'] = [int(x) for x in row[6].split(',')[:-1]]\n",
    "            gene_dict[row[7]][row[0][:15]]['length'] = mRNA_len(gene_dict[row[7]][row[0][:15]]['starts'],\n",
    "                                                               gene_dict[row[7]][row[0][:15]]['ends'])\n",
    "            if int(row[3]) < gene_dict[row[7]]['global_start']:\n",
    "                gene_dict[row[7]]['global_start'] = int(row[3])\n",
    "            if int(row[4]) > gene_dict[row[7]]['global_end']:\n",
    "                gene_dict[row[7]]['global_end'] = int(row[4])\n",
    "\n",
    "        try:\n",
    "            gene_dict[row[7]]['labels'] = np.zeros(\n",
    "                (int(gene_dict[row[7]]['global_end']) - \n",
    "                 int(gene_dict[row[7]]['global_start'])))\n",
    "        except ValueError:\n",
    "            print(row[7])\n",
    "            print(gene_dict[row[7]]['global_start'], gene_dict[row[7]]['global_end'])\n",
    "\n",
    "print(\"Took {} seconds to construct the gene dict\".format(time.time() - start_time))\n",
    "print(\"Number of genes in the region of interest:\", len(gene_dict.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import hg38\n",
    "\n",
    "Loading human genome (version hg38) to extract the respective \"extended\" gene sequences (covering all the transcripts of the gene + flanking ends of 1000nt for context)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to download and unpack the genome file into the ./data directory:\n",
    "# wget --timestamping 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz' -O hg38.fa.gz\n",
    "# gunzip hg38.fa.gz\n",
    "path_to_file = './data/hg38.fa'\n",
    "\n",
    "hg38 = {}\n",
    "\n",
    "with open(path_to_file, mode='r') as handle:\n",
    "\n",
    "    for record in SeqIO.parse(handle, 'fasta'):\n",
    "\n",
    "        identifier = record.id\n",
    "        description = record.description\n",
    "        sequence = record.seq\n",
    "        \n",
    "        hg38[identifier] = sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import count files by tissue type + AML\n",
    "\n",
    "## construct the library separately for each tissue type and count file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_library(gene_dict, counts):\n",
    "    library = {}\n",
    "    start_time = time.time()\n",
    "    \n",
    "    for sample in counts.index:\n",
    "        library[sample] = {}\n",
    "        for gene in gene_dict.keys():\n",
    "            for tr in gene_dict[gene].keys():\n",
    "                \n",
    "                if tr in counts.columns:\n",
    "                    if gene not in library[sample].keys():\n",
    "                        library[sample][gene] = {}\n",
    "                        library[sample][gene] = {}\n",
    "                        length = int(gene_dict[gene]['global_end']) - int(gene_dict[gene]['global_start']) + 1\n",
    "                        library[sample][gene]['alabels'] = np.zeros(length)\n",
    "                        library[sample][gene]['dlabels'] = np.zeros(length)\n",
    "                        library[sample][gene]['norm_factor'] = 0\n",
    "                        gs = int(gene_dict[gene]['global_start'])\n",
    "                        ge = int(gene_dict[gene]['global_end'])\n",
    "                        l = gene_dict[gene][tr]['length']\n",
    "                        for s in gene_dict[gene][tr]['starts']:\n",
    "                            # normalise to tpkm\n",
    "                            library[sample][gene]['alabels'][int(s)-gs] += float(counts.at[sample,tr]/l)\n",
    "                        for s in gene_dict[gene][tr]['ends']:\n",
    "                            # normalise to tpkm\n",
    "                            library[sample][gene]['dlabels'][int(s)-gs] += float(counts.at[sample,tr]/l)\n",
    "                        # normalise to tpkm\n",
    "                        library[sample][gene]['norm_factor'] += float(counts.at[sample,tr]/l)\n",
    "                    else:\n",
    "                        gs = int(gene_dict[gene]['global_start'])\n",
    "                        ge = int(gene_dict[gene]['global_end'])\n",
    "                        l = gene_dict[gene][tr]['length']\n",
    "                        for s in gene_dict[gene][tr]['starts']:\n",
    "                            # normalise to tpkm\n",
    "                            library[sample][gene]['alabels'][int(s)-gs] += float(counts.at[sample,tr]/l)\n",
    "                        for s in gene_dict[gene][tr]['ends']:\n",
    "                            # normalise to tpkm\n",
    "                            library[sample][gene]['dlabels'][int(s)-gs] += float(counts.at[sample,tr]/l)\n",
    "                        # normalise to tpkm\n",
    "                        library[sample][gene]['norm_factor'] += float(counts.at[sample,tr]/l)\n",
    "                        \n",
    "    print(\"All samples calc in {} seconds\".format(time.time() - start_time))\n",
    "    return library\n",
    "\n",
    "\n",
    "def labels_squeeze(labels, c):\n",
    "    ind = np.nonzero(labels)[0]\n",
    "    if c==0:\n",
    "        c = 1\n",
    "    labels = [round(labels[i]/c, 4) for i in ind]\n",
    "    labels_ = ''\n",
    "    for i, l in zip(ind,labels):\n",
    "        labels_ += str(i) + ' ' + str(l) + ',' \n",
    "    return labels_\n",
    "\n",
    "\n",
    "def exons(gene):\n",
    "    exons = ''\n",
    "    gs = int(gene['global_start'])\n",
    "    for key in gene.keys():\n",
    "        if 'ENST' in key:\n",
    "            starts = gene[key]['starts']\n",
    "            ends = gene[key]['ends']\n",
    "            for x in zip(starts, ends):\n",
    "                exons += str(x[0]-gs) + ' ' + str(x[1]-gs) + ','\n",
    "            if exons:\n",
    "                exons = exons[:-1] + ';'\n",
    "    return exons\n",
    "\n",
    "\n",
    "def save_labels_jsonl(counts, gene_dict, hg38):\n",
    "    \n",
    "    if not os.path.exists('./data/main_inputs/'):\n",
    "        os.makedirs('./data/main_inputs/')\n",
    "    \n",
    "    # normalise to cpm first\n",
    "    F = counts.sum(axis = 1)/10**6\n",
    "    counts = counts.divide(F, axis='index')\n",
    "    \n",
    "    context = 1000\n",
    "\n",
    "    start_time = time.time()\n",
    "    library = construct_library(gene_dict, counts)\n",
    "    for sample in library.keys():\n",
    "        print(sample)\n",
    "        with open('./data/main_inputs/'+sample+'.jsonl', 'a') as f1:\n",
    "            # genes\n",
    "            for gene in library[sample].keys():\n",
    "                # header: gene name \n",
    "                #header = '@' + gene\n",
    "                #f1.write(header + os.linesep)\n",
    "                jsonl_dict = {}\n",
    "\n",
    "                gs = int(gene_dict[gene]['global_start'])\n",
    "                ge = int(gene_dict[gene]['global_end'])\n",
    "                seq = str(hg38[gene_dict[gene]['chr']][gs - context : ge + context + 1])\n",
    "                # next line: sequence + context flanks on each side\n",
    "                #f1.write(seq + os.linesep)\n",
    "                jsonl_dict[gene] = seq\n",
    "\n",
    "                # next two lines: labels, acceptor then donor, length = seq length - flanks\n",
    "                c = library[sample][gene]['norm_factor']\n",
    "                #f1.write(labels_squeeze(library[sample][gene]['alabels'], c) + os.linesep)\n",
    "                #f1.write(labels_squeeze(library[sample][gene]['dlabels'], c) + os.linesep)\n",
    "                jsonl_dict['alabels'] = labels_squeeze(library[sample][gene]['alabels'], c)\n",
    "                jsonl_dict['dlabels'] = labels_squeeze(library[sample][gene]['dlabels'], c)\n",
    "                jsonl_dict['exons'] = exons(gene_dict[gene])\n",
    "                \n",
    "                json.dump(jsonl_dict, f1)\n",
    "                f1.write('\\n')\n",
    "                \n",
    "    print(\"All samples saved in {} seconds\".format(time.time() - start_time))\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.DS_Store', 'AML.csv', 'adipose.csv', 'blood.csv', 'brain.csv', 'breast.csv', 'colon.csv', 'heart.csv', 'kidney.csv', 'liver.csv', 'lung.csv', 'lymph.csv', 'prostate.csv', 'skeletal.csv', 'testis.csv', 'thyroid.csv']\n",
      "AML.csv ; shape: (1874, 178136)\n",
      "All samples calc in 1.5012640953063965 seconds\n",
      "AML1\n",
      "AML2\n",
      "AML3\n",
      "All samples saved in 3.7682077884674072 seconds\n",
      "adipose.csv ; shape: (200, 178136)\n",
      "All samples calc in 1.005049705505371 seconds\n",
      "ADP1\n",
      "ADP2\n",
      "ADP3\n",
      "All samples saved in 2.6345481872558594 seconds\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "path = './data/counts_by_tissue/'\n",
    "dirs = os.listdir(path)\n",
    "print(dirs)\n",
    "\n",
    "for file in dirs:\n",
    "    if 'csv' in file:\n",
    "        counts = pd.read_csv(path+file, sep=',', header=0, index_col=0, low_memory=False)\n",
    "        print(file, '; shape:', np.shape(counts))\n",
    "        save_labels_jsonl(counts, gene_dict, hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
